dev/PushNotifications/PushNotificationManager.cpp

```diff
@@ -0,0 +1,182 @@
+
+        uint8_t retryCount = 0;
+        winrt::hresult channelRequestResult = E_PENDING;
+        winrt::Microsoft::Windows::PushNotifications::PushNotificationChannelStatus status = PushNotificationChannelStatus::InProgress;
+
+        winrt::Microsoft::Windows::PushNotifications::PushNotificationCreateChannelStatus
+            channelStatus = { channelRequestResult, status, retryCount };
+
+        progress(channelStatus);
+
+        PushNotificationChannelManager channelManager{};
+        winrt::Windows::Networking::PushNotifications::PushNotificationChannel pushChannelReceived{ nullptr };
+
+        for (auto backOffTimeInSeconds = c_minBackoffSeconds; backOffTimeInSeconds <= c_maxBackoffSeconds * 2; backOffTimeInSeconds *= 2)
+        {
+            try
+            {
+                pushChannelReceived = co_await channelManager.CreatePushNotificationChannelForApplicationAsync();
+                channelRequestResult = S_OK;
+                status = PushNotificationChannelStatus::CompletedSuccess;
+                break;
+            }
+            catch (...)
+            {
+
+                auto channelRequestException = hresult_error(to_hresult(), take_ownership_from_abi);
+
+                if ((backOffTimeInSeconds <= c_maxBackoffSeconds) && isChannelRequestRetryable(channelRequestException.code()))
+                {
+                    channelStatus.extendedError = channelRequestException.code();
+                    channelStatus.status = PushNotificationChannelStatus::InProgressRetry;
+                    channelStatus.retryCount = ++retryCount;
+
+                    progress(channelStatus);
+                }
+                else
+                {
+                    channelRequestResult = channelRequestException.code();
+                    status = PushNotificationChannelStatus::CompletedFailure;
+                    break;
+                }
+            }
+
+            co_await winrt::resume_after(std::chrono::seconds(backOffTimeInSeconds));
+        }
+
+        
+        if (status == PushNotificationChannelStatus::CompletedSuccess)
+        {
+            // Returns a com_ptr to the implementation type
+            auto pushChannel =
+                winrt::make_self<winrt::Microsoft::Windows::PushNotifications::implementation::PushNotificationChannel>(pushChannelReceived);
+
+            channelResult = winrt::make<winrt::Microsoft::Windows::PushNotifications::implementation::PushNotificationCreateChannelResult>(
+                *(pushChannel.get()),
+                channelRequestResult,
+                status);
+        }
+        else if(status == PushNotificationChannelStatus::CompletedFailure)
+        {
+            channelResult = winrt::make<winrt::Microsoft::Windows::PushNotifications::implementation::PushNotificationCreateChannelResult>(
+                nullptr,
+                channelRequestResult,
+                status);
+        }
+
+        co_return channelResult;
+    }
```

Comment:
I assume the progress token is automatically updated when you return from the function, is that correct?

Reply:
whenever you call progress(); function the progress status is updated. When you return from the function depending on the scenario we are at progress API may or may not be called. Thanks.